src/ray/raylet/node_manager.cc

```diff
@@ -1081,58 +1082,49 @@ void NodeManager::NodeRemoved(const NodeID &node_id) {
   // Notify the object directory that the node has been removed so that it
   // can remove it from any cached locations.
   object_directory_->HandleNodeRemoved(node_id);
-
-  // Clean up workers that were owned by processes that were on the failed
-  // node.
-  rpc::WorkerDeltaData data;
-  data.set_raylet_id(node_id.Binary());
-  HandleUnexpectedWorkerFailure(data);
 }
 
-void NodeManager::HandleUnexpectedWorkerFailure(const rpc::WorkerDeltaData &data) {
-  const WorkerID worker_id = WorkerID::FromBinary(data.worker_id());
-  const NodeID node_id = NodeID::FromBinary(data.raylet_id());
-  if (!worker_id.IsNil()) {
-    RAY_LOG(DEBUG).WithField(worker_id) << "Worker failed";
-    failed_workers_cache_.insert(worker_id);
-  } else {
-    RAY_CHECK(!node_id.IsNil());
-    failed_nodes_cache_.insert(node_id);
+void NodeManager::HandleNodeRemoved(const NodeID &node_id) {
+  RAY_CHECK(!node_id.IsNil());
+  failed_nodes_cache_.insert(node_id);
+
+  cluster_task_manager_->CancelAllTasksOwnedBy(node_id);
+
+  for (const auto &pair : leased_workers_) {
+    auto &worker = pair.second;
+    const auto owner_node_id = NodeID::FromBinary(worker->GetOwnerAddress().raylet_id());
+    RAY_CHECK(!owner_node_id.IsNil());
+    if (worker->IsDetachedActor() || owner_node_id != node_id) {
+      continue;
+    }
+    // If the leased worker's owner was on the failed node, then kill the leased
+    // worker.
+    RAY_LOG(INFO).WithField(worker->WorkerId()).WithField(owner_node_id)
+        << "The leased worker is killed because the owner node died.";
+    KillWorker(worker);
   }
+}
+
+void NodeManager::HandleUnexpectedWorkerFailure(const WorkerID &worker_id) {
+  RAY_CHECK(!worker_id.IsNil());
+  RAY_LOG(DEBUG).WithField(worker_id) << "Worker failed";
+  failed_workers_cache_.insert(worker_id);
 
-  cluster_task_manager_->CancelAllTaskOwnedBy(worker_id);
```

Comment:
We used to only cancel tasks when the owner worker fails but not when the owner node fails.